pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,356 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/cel-go/common/types/ref"
+
+	runtime2 "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// combinedStatusResolution is a struct that represents the resolution of a
+// combinedstatus. A combinedstatus resolution is associated with a (binding,
+// object) tuple.
+// A resolution's logic is as follows:
+// - Composition:
+//   - A combinedstatus resolution is composed of multiple statuscollector
+//     data.
+//   - statuscollector data is composed of collector-clauses. Each
+//     collector-clause has an expression that is evaluated against every
+//     workstatus that is associated with the statuscollector, and the result
+//     is cached.
+//
+// - Flow:
+//   - When a statuscollector is added to the combinedstatus resolution,
+//     every clause in the statuscollector is evaluated against every workstatus
+//     that is associated with the statuscollector.
+//   - When a workstatus is added to / modified in the combinedstatus
+//     resolution, every clause in every statuscollector is evaluated against
+//     the workstatus. The caching of collector-clause -> evaluation makes
+//     calculating combinedFields efficient if present.
+type combinedStatusResolution struct {
+	sync.RWMutex
+
+	// statusCollectorNameToData is a map of status collector names to
+	// their corresponding data.
+	// If a status collector name is mapped to nil, it means that the status
+	// collector has not been processed.
+	statusCollectorNameToData map[string]*statusCollectorData
+	// collectionDestinations is a set of destinations that are expected to be
+	// collected from.
+	collectionDestinations sets.Set[string]
+}
+
+// statusCollectorData is a struct that represents the data of a status
+// collector in a combinedstatus resolution.
+type statusCollectorData struct {
+	*v1alpha1.StatusCollectorSpec
+
+	// workStatusRefToData is a map of workstatus-hosting WEC name to the
+	// evaluation of the workstatus against the statuscollector's clauses.
+	workStatusRefToData map[string]*workStatusData
```

Comment:
This only holds an entry if it passed the collector's filter, right? If so then that should be stated in the comment on the field.